util/fipstools/inject_hash/macho_parser/macho_parser.c

```diff
@@ -0,0 +1,218 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <stdint.h>
+
+#include "common.h"
+#include "macho_parser.h"
+
+#define TEXT_INDEX 0
+#define CONST_INDEX 1
+#define SYMTABLE_INDEX 2
+#define STRTABLE_INDEX 3
+
+// Documentation for the Mach-O structs can be found in macho-o/loader.h and mach-o/nlist.h
+int read_macho_file(const char *filename, machofile *macho) {
+    FILE *file = NULL;
+    struct load_command *load_commands = NULL;
+    uint32_t bytes_read;
+    int ret = 0;
+
+    file = fopen(filename, "rb");
+    if (file == NULL) {
+        LOG_ERROR("Error opening file %s", filename);
+        goto end;
+    }
+
+    bytes_read = fread(&macho->macho_header, 1, sizeof(struct mach_header_64), file);
+    if (bytes_read != sizeof(struct mach_header_64)) {
+        LOG_ERROR("Error reading macho_header from file %s", filename);
+        goto end;
+    }
+    if (macho->macho_header.magic != MH_MAGIC_64) {
+        LOG_ERROR("File is not a 64-bit Mach-O file");
+        goto end;
+    }
+
+    load_commands = malloc(macho->macho_header.sizeofcmds);
+    if (load_commands == NULL) {
+        LOG_ERROR("Error allocating memory for load_commands");
+        goto end;
+    }
+    bytes_read = fread(load_commands, 1, macho->macho_header.sizeofcmds, file);
+    if (bytes_read != macho->macho_header.sizeofcmds) {
+        LOG_ERROR("Error reading load commands from file %s", filename);
+        goto end;
+    }
+
+    // We're only looking for __text, __const in the __TEXT segment, and the string & symbol tables
+    macho->num_sections = 4;
+    macho->sections = malloc(macho->num_sections * sizeof(section_info));
+    if (macho->sections == NULL) {
+        LOG_ERROR("Error allocating memory for macho sections");
+    }
+
+    int text_found = 0;
+    int const_found = 0;
+    int symtab_found = 0;
+
+    for (uint32_t i = 0; i < macho->macho_header.sizeofcmds / sizeof(struct load_command); i += load_commands[i].cmdsize / sizeof(struct load_command)) {
```

Comment:
This logic looks suspcious. Is `load_commands[i].cmdsize` not always the same as `sizeof(struct load_command)`?  This logic assumes that it can be an exact multiple (>1). A comment here should explain the "why".

Otherwise, if the two values are equal it could be simplified:
```
    const uint32_t num_cmds = macho->macho_header.sizeofcmds / sizeof(struct load_command);
    for (uint32_t i = 0; i < num_cmds; i += 1) {
```

Reply:
You're right, that was an oversight on my part - thanks for pointing it out!

Reply:
The reason this works is that sizeof(struct load_command) results in 8 bytes (the struct consists of two `uint32_t`s). The documentation in `loader.h` says that `sizeofcmds` actually needs to be a multiple of 8 on 64bit systems, so this should work here.

I'll add a comment and your simplified code.